src/order/heyting_algebra.lean

```diff
@@ -0,0 +1,136 @@
+/-
+Copyright (c) 2020 Chris A. Upshaw. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris A. Upshaw.
+
+Type class hierarchy for Heyting algebras.
+-/
+import order.bounded_lattice
+
+set_option old_structure_cmd true
+
+/-!
+# Heyting algebra
+
+In this file we introduce Heyting algebras.
+
+A Heyting algebra is a bounded distributed lattice with an internal implication.
+
+Having an internal implication and a bottom allows for defining an (possilbe intuitionistic)
+complement.
+
+## Notation
+
+ - ` ⇨ ` : implication as an operator internal to a lattice.
+
+-/
+
+universes u v
+variables {α : Type u} {w x y z : α}
+
+class has_internal_imp (α : Type*) :=
+  (imp : α → α → α)
+
+infixr ` ⇨ `:60 := has_internal_imp.imp
+
+/-- A heyting algebra is a bounded distributive lattice with a internal implication right adjoint to
+ the meet. This also allows the definition of a psudo-complement `ᶜ` such that `x ⊓ xᶜ = ⊥`,
+ (but not nessarily `x ⊔ xᶜ = ⊤`).
+
+  This is a generalization of the logic of propositions and of open sets of a topology, though
+  heyting algebra morphisms are distinct from images of continous functions.
+  -/
+class heyting_algebra α extends bounded_distrib_lattice α, has_internal_imp α :=
```

Comment:
I think it would be nice to have a constructor for Heyting algebras which assumes `bounded_lattice` (not distrib) and the axiom here - since the axiom implies distributivity.